src/libraries/System.Security.Cryptography.Cose/tests/CoseMessageTests.DecodeSign1.cs

```diff
@@ -69,5 +69,30 @@ public void DecodeSign1_IncorrectStructure()
             writer.WriteEndArray();
             Assert.Throws<CryptographicException>(() => CoseMessage.DecodeSign1(writer.Encode()));
         }
+
+        [Fact]
+        public void DecodeSign1_IndefiniteLengthArray()
+        {
+            byte[] cborPayload = ByteUtils.HexToByteArray("D29F43A10126A10442313154546869732069732074686520636F6E74656E742E58408EB33E4CA31D1C465AB05AAC34CC6B23D58FEF5C083106C4D25A91AEF0B0117E2AF9A291AA32E14AB834DC56ED2A223444547E01F11D3B0916E5A4C345CACB36FF");
+            CoseSign1Message msg = CoseMessage.DecodeSign1(cborPayload);
+
+            Assert.True(msg.VerifyEmbedded(DefaultKey));
+        }
+
+        [Fact]
+        public void DecodeSign1_IndefiniteLengthArray_MissingBreak()
+        {
+            byte[] cborPayload = ByteUtils.HexToByteArray("D29F43A10126A10442313154546869732069732074686520636F6E74656E742E58408EB33E4CA31D1C465AB05AAC34CC6B23D58FEF5C083106C4D25A91AEF0B0117E2AF9A291AA32E14AB834DC56ED2A223444547E01F11D3B0916E5A4C345CACB36");
+            CryptographicException ex = Assert.Throws<CryptographicException>(() => CoseMessage.DecodeSign1(cborPayload));
+            Assert.IsType<CborContentException>(ex.InnerException);
+        }
+
+        [Fact]
+        public void DecodeSign1_IndefiniteLengthArray_LargerByOne()
+        {
+            byte[] cborPayload = ByteUtils.HexToByteArray("D29F43A10126A10442313154546869732069732074686520636F6E74656E742E58408EB33E4CA31D1C465AB05AAC34CC6B23D58FEF5C083106C4D25A91AEF0B0117E2AF9A291AA32E14AB834DC56ED2A223444547E01F11D3B0916E5A4C345CACB3640FF");
+            CryptographicException ex = Assert.Throws<CryptographicException>(() => CoseMessage.DecodeSign1(cborPayload));
+            Assert.IsType<InvalidOperationException>(ex.InnerException);
```

Comment:
It feels to me like this exception should look the same as the definite array case.

Definite:

```
System.Security.Cryptography.CryptographicException: Error while decoding COSE message. COSE_Sign1 must be an array of four elements.
```

Indefinite:

```
System.Security.Cryptography.CryptographicException: Error while decoding COSE message. See the inner exception for details.
----- Inner Exception -----
  System.InvalidOperationException: Not at end of the indefinite-length data item.
```

And for the "one too short case" (which could use a test) we get the same informative exception with the definite length, but indefinite becomes

```
System.Security.Cryptography.CryptographicException: Error while decoding COSE message. See the inner exception for details.
----- Inner Exception -----
  System.InvalidOperationException: Cannot perform the requested operation, the next CBOR data item is of major type '7'
```

(I think '7' is what'll report, at least).

Reply:
Similar concern as your previous comment https://github.com/dotnet/runtime/pull/74215#discussion_r953136556.
Is it worth to check up-front indefinite-length arrays length just to get nice exceptions?

Reply:
There are a couple of potential reasons someone gets here, but one of them is getting the method call backwards between COSE_Sign1 and COSE_Sign formats.  In the case of definite length arrays there's a "you might be the wrong item" message.  For indefinite the message is way more obscure.

Assuming the developer trying to read a file is not the one who wrote it, they don't control the definite or indefinite encoding; so it seems wrong to punish them for being in a corner case.  So, yeah, I think we should check the length up front.

I agree that the only way to really pre-count it is to use ReadEncodedValue() and make a second reader (and, for indefinite, a third one), but ideally we won't do that when it's a definite length.  So my recommendation is probably

* Use pre-counting now.
* Noodle around in the CBOR project to make sure that adding a PeekEncodedValue() member is as easy as I think it should be
* API proposal for PeekEncodedValue()
* When implementing PeekEncodedValue() also come here and make use of it:

```C#
ReadOnlyMemory<byte> encodedArray = reader.PeekEncodedValue();
int? len = reader.ReadStartArray();

if (!len.HasValue)
{
    // make a reader over encodedArray, count how many items were in it.
}

if (len != TheAppropriateConstant)
{
    throw new ...
}
```

That makes definite arrays continue as they are (with a minor hit for double-reading the start array and saving the ReadOnlyMemory), and only indefinite arrays pay the allocation of a second reader, and there's never a third reader.